rts/Benchmarking.md

```diff
@@ -0,0 +1,145 @@
+
+- `total_pages_in_use` is **Wasm** pages, not host pages!
+
+drun can be built with the latest stable Rust toolchain (1.56 as I write this)
+by just navigating into `rs/drun` and running `cargo build --release`. No need
+for nix. The executable will be in `rs/target/release/drun`.
+
+Copy this executable to your `$PATH` for ease of access.
+
+## Finding a canister
+
+So far the only canister I have that is suitable for benchmarks is
+https://github.com/dfinity/cancan-archived. I'm using commit 302343a.
+
+Run `vessel sources` to get `--package` flags for the dependencies:
+
+```
+$ vessel sources
+--package base .vessel/base/dfx-0.7.0-beta.2/src --package crud .vessel/crud/master/src --package sequence .vessel/sequence/master/src
+```
+
+Note that the base version above is buggy and you need the master branch, so
+clone motoko-base and update the path above with the path to your clone.
+
+
+```
+<path to moc> <package flags> service/CanCan.mo -o <wasm file name> --force-gc <gc strategy>
+```
+
+- `<path to moc>`: Path to `moc` executable
+- `<package flags>`: Flags printed by the vessel command shown above. Make sure
+  to update base path to point at motoko-base master.
+- `<wasm file name>`: Name of the generated .wasm file. Make sure to add some
+  prefix/suffixes to this binary to describe the compile flags. E.g.
+  `cancan_copying_gc_my_awesome_patch_tweaked_scheduling_wow.wasm`.
+- `<gc strategy>`: `--copying-gc` or `--compacting-gc`.
+
+## Writing a drun script
+
+Now that we have a canister, we need a drun script to install the canister, and
+send messages to it. The resulting `canister_perf.csv` file will have one line
+for each `ingress` line in the drun script.
+
+For the CanCan backend canister above, I use a script like this:
+
+```
+create
+install rwlgt-iiaaa-aaaaa-aaaaa-cai <canister Wasm path> ""
+ingress rwlgt-iiaaa-aaaaa-aaaaa-cai createProfile "DIDL\x02n\x01m{\x02q\x00\x05test0\x00"
+ingress rwlgt-iiaaa-aaaaa-aaaaa-cai createProfile "DIDL\x02n\x01m{\x02q\x00\x05test1\x00"
+ingress rwlgt-iiaaa-aaaaa-aaaaa-cai createProfile "DIDL\x02n\x01m{\x02q\x00\x05test2\x00"
+...
+```
+
+This calls `createProfile` a few hundred times.
+
+Now I save this as `test.drun` run drun:
+
+```
+drun test.drun
+```
+
+After that I have `canister_perf.csv` file with the results. Make sure to copy
+this CSV file because if you accidentally run drun again (with up-arrow + enter
+or something like that) it will destroy your results.
+
+## Comparing results
+
+- Compile the canister with different flags, with different output file names.
+- Run the drun script twice, updating the canister Wasm path in the drun script
+  before running drun.
+- Copy generated .csv file after each run, give it a descriptive name.
+
+After that you will have csv files that you can compare.
+
+## How to compare CSV files
+
+Most of the time I simply do `nvim -d <csv1> <csv2>`.
+
+I have a Rust program to generate plots for instructions, dirtied pages etc.
```

Comment:
```suggestion
I have a Rust program to generates plots for instructions, dirtied pages etc.
```

Reply:
Oh, this sounds interesting...

Reply:
https://github.com/osa1/generate_drun_bench_plots